### ISSUE TITLE
Object oriented gravity alignment

### ISSUE THREAD

**Issue by marcoreni** [imgproxy/imgproxy#1381]
We're using smart gravity / object oriented gravity with excellent results, but we have some specific scenarios where we'd like to be able to specify an "alignment" of the resulting image.

For example. We have the following starting image:
![original](https://github.com/user-attachments/assets/6f3f2ada-eeb3-446c-8ad1-f79e6062002c)

The person in this photo is aligned to the right for business reason. We'd like to be able to generate a 768x480 image where we push the person "as right as possible", so something like this (we highlighted the person for clarity):

<img width="708" alt="image" src="https://github.com/user-attachments/assets/0fe73b24-8472-410e-b30b-00de81d7fcea">

We tried using the following parameters: `gravity:objw:face:2/rs:fill-down:767:480:1`, but the result has some "space" on the right (obviously because the face is put at the center of the image):

![image](https://github.com/user-attachments/assets/ce7cf1c5-d891-45fb-9583-45ca1fc18af7)

Assuming that we use an ML model that recognizes persons, instead of faces, would there be a way to "move" the image so that the recognized object is not at the center but aligned to the right?

**Comment by DarthSim** [comment]
Hey @marcoreni!

We're planning to extend the object-oriented crop functionality in the next release. I'll add this feature to its roadmap.

**Comment by marcoreni** [comment]
Hey @DarthSim ! Thanks for the quick reply! 

Is there an hypthotetical timeframe for the next release?

**Comment by DarthSim** [comment]
I'm not sure about the release, but I guess I can show you something to test in the `latest` build sometime next week.

**Comment by marcoreni** [comment]
Thanks! Have a great weekend!

**Comment by DarthSim** [comment]
Hey @marcoreni!

I'm about to start implementing this feature, and I would like to hear your opinion.

I see two ways to reach your goal:

1. I will add an `objects_position` processing option to control how imgproxy should position objects inside the crop area when the objects' area width/height is less than the crop area width/height. That option will accept the same values as `gravity` except the "smart" ones. The downside is that you must specify this option for each case.

2. I will add an `IMGPROXY_OBJECT_DETECTION_CROP_POSITIONING` config and the corresponding processing option. They will accept two values:
  - `center`—the current behavior. If the width/height of the objects' area is less than the width/height of the crop area, imgproxy will try to position the objects as close to the center of the crop area as possible.
  - `proportional`—imgproxy will try to set objects' offsets inside the crop area proportional to their offsets in the whole image area. Here's an example that uses your image:
  ![obj-pos-prop](https://github.com/user-attachments/assets/1285f00d-17b5-43e1-a23f-9a3de63ab02e)
  As you see, there's still an offset on the right, but it is decreased proportionally to the left offset. This is not exactly what you described, but it maintains the scene of the photo.

  - `minimal_shift`—imgproxy shifts the crop area from the center of the image as little as possible so it has the best total score. Here's an example with your image:
  ![obj-pos-min-shift](https://github.com/user-attachments/assets/3f26fe6c-e5b6-479e-9005-dd7bad87c271)
  This mode produces the result you've described, but I highly doubt it will be handy for other cases.

I don't see how these two ways could work together, so we must choose one.

**Comment by DarthSim** [comment]
Actually, I already implemented a POC for the 2nd way to prepare those examples for you, so if it works for you, I can quickly polish it and push it to the `latest` build 😅

**Comment by marcoreni** [comment]
Ciao @DarthSim ,

I think the solution you described it's good, but I need a bit of clarification on the `minimal_shift` behavior. What do you mean by "total score"? You mean that the image will be shifted until a "further shift" would lower the AI recognition score? In this scenario, how would it decide that the image should go left/right/top/bottom? 

Would it be able to "change the proportions" of an image? For example, in the following image the person is at the center. Would the `minimal_shift` be able to move the person to the right?

![image](https://github.com/user-attachments/assets/fe18ee89-c0d4-4938-81e3-4ed77031e934)

**Comment by DarthSim** [comment]
Oh, now I get it. You want to move the man to the right not because he is on the right in the original image but because you want him on the right in the resulting image. If that's the case, all the "automated" solutions won't help, and we should go the 1st way.

**Comment by marcoreni** [comment]
Yes, the scenario is that we need the relevant part of the image to be "on the right" of the page, but in some cases the asset we receive from third parties is not appropriate to our need, so being able to "shift the center" would be ideal.

**Comment by DarthSim** [comment]
Hey @marcoreni!

It's still Friday in my place, so technically, I kept my promise 😅

Please pull the `latest` build and add `op:ea` to your processing options. I haven't added this to the documentation yet, but long story short:

I added the `objects_position`/`obj_pos`/`op` processing option, which controls the position of detected objects inside the crop area. It accepts the same values as the `gravity` option except for `sm`, `obj`, and `obj`. Also, it accepts the `prop` value which does the thing I described before: imgproxy will try to set objects' offsets inside the crop area proportional to their offsets in the whole image area. Since you need to move objects to the right edge, you need the `op:ea` option. If you want to add a little offset so the object doesn't stick right to the edge, you can add this the same way as when for the `gravity option`: `op:ea:0.02`.

**Comment by marcoreni** [comment]
Ciao @DarthSim , still friday here as well! 

It looks perfect, it covers all of our scenarios and some further scenarios that we _hoped to cover but thought were too exoteric_ 😄 .

Many thanks!
